moz_security/sandboxes/heka/analysis/moz_security_webinj.lua

```diff
@@ -0,0 +1,164 @@
+-- xss_violation_type = "fxa:webinj_xss" -- required in violations mode, iprepd violation type
+-- sqli_violation_type = "fxa:webinj_sqli" -- required in violations mode, iprepd violation type
+```
+--]]
+
+require "string"
+require "table"
+
+libi    = require "libinjection"
+uri     = require "lpeg.uri"
+
+local id_field          = read_config("id_field") or error("id_field must be configured")
+local id_fieldc         = read_config("id_field_capture")
+local request_field     = read_config("request_field") or error("request_field must be configured")
+local request_fieldc    = read_config("request_field_capture")
+local list_max_size     = read_config("list_max_size") or 500
+local strip_nul         = read_config("strip_nul")
+
+local tbsend
+local violation_type
+local send_iprepd = read_config("send_iprepd")
+if send_iprepd then
+    tbsend = require "heka.iprepd".send
+    xss_violation_type = read_config("xss_violation_type") or error("xss_violation_type must be configured")
+    sqli_violation_type = read_config("sqli_violation_type") or error("sqli_violation_type must be configured")
+end
+
+local list = {}
+local list_size = 0
+
+function process_message()
+    if list_size >= list_max_size then
+        return 0 -- list full, stop analyzing until the next timer interval
+    end
+
+    local id = read_message(id_field)
+    if not id then return -1, "no id_field" end
+    if id_fieldc then
+        id = string.match(id, id_fieldc)
+        if not id then return 0 end -- no error as the capture may intentionally reject entries
+    end
+    local orig_req = read_message(request_field)
+    if not orig_req then return -1, "no request_field" end
+    local req = orig_req
+    if request_fieldc then
+        req = string.match(req, request_fieldc)
+        if not req then return 0 end -- no error as the capture may intentionally reject entries
+    end
+    if strip_nul then req = string.gsub(req, "%%00", "") end
+
+    local m = uri.uri_reference:match(req)
+    if not m or not m.query then return 0 end
+    local params = uri.url_query:match(m.query)
+    if not params then return 0 end -- don't treat as an error, request may not have parameters
+
+    local foundtype = nil
+    for k,v in pairs(params) do
+        if libi.xss(v, string.len(v)) == 1 then
+            foundtype = "xss"
+        else
+            local state = libi.sqli_state()
+            if not state then return -1, "error allocating sqli state" end
+            libi.sqli_init(state, v, string.len(v), 0)
+            if libi.is_sqli(state) == 1 then
+                foundtype = "sqli"
+            end
+        end
+    end
+
+    if foundtype then
+        if not list[id] then
+            list[id] = {
+                cnt         = 0,
+                lastsqli    = nil,
+                lastxss     = nil,
+                sqli        = false,
+                xss         = false
```

Comment:
The xss and sqli flags seem redundant since you can just the the last* value itself